velox/dwio/dwrf/common/FileMetadata.cpp

```diff
@@ -0,0 +1,254 @@
+    : offset_{info.offset()},
+      indexLength_{info.indexlength()},
+      dataLength_{info.datalength()},
+      footerLength_{info.footerlength()},
+      numberOfRows_{info.numberofrows()},
+      encryptStripeId_{
+          info.has_encryptstripeid()
+              ? std::optional<uint64_t>(info.encryptstripeid())
+              : std::nullopt} {
+  for (int32_t i = 0; i < info.encryptedlocalkeys_size(); ++i) {
+    encryptedLocalKeys_.push_back(info.encryptedlocalkeys((i)));
+  }
+}
+
+PostScript::PostScript(
+    uint64_t footerLength,
+    dwio::common::CompressionKind compression,
+    uint64_t compressionBlockSize,
+    uint32_t writerVersion)
+    : footerLength_{footerLength},
+      compression_{compression},
+      compressionBlockSize_{compressionBlockSize},
+      writerVersion_{static_cast<WriterVersion>(writerVersion)} {}
+
+PostScript::PostScript(const proto::PostScript& ps)
+    : footerLength_{ps.footerlength()},
+      compression_{
+          ps.has_compression()
+              ? static_cast<dwio::common::CompressionKind>(ps.compression())
+              : dwio::common::CompressionKind::CompressionKind_NONE},
+      compressionBlockSize_{
+          ps.has_compressionblocksize()
+              ? ps.compressionblocksize()
+              : dwio::common::DEFAULT_COMPRESSION_BLOCK_SIZE},
+      writerVersion_{
+          ps.has_writerversion()
+              ? static_cast<WriterVersion>(ps.writerversion())
+              : WriterVersion::ORIGINAL},
+      cacheMode_{static_cast<StripeCacheMode>(ps.cachemode())},
+      cacheSize_{ps.cachesize()} {}
+
+PostScript::PostScript(const proto::orc::PostScript& ps)
+    : fileFormat_{dwio::common::FileFormat::ORC},
+      footerLength_{ps.footerlength()},
+      compression_{detail::orcCompressionToCompressionKind(ps.compression())},
+      compressionBlockSize_{ps.compressionblocksize()},
+      writerVersion_{static_cast<WriterVersion>(ps.writerversion())},
+      metadataLength_{ps.metadatalength()},
+      stripeStatisticsLength_{ps.stripestatisticslength()} {}
+
+Footer::Footer(proto::Footer* footer)
+    : headerLength_{footer->has_headerlength() ? footer->headerlength() : 0},
+      contentLength_{footer->has_contentlength() ? footer->contentlength() : 0},
+      numberOfRows_{footer->has_numberofrows() ? footer->numberofrows() : 0},
+      rowIndexStride_{
+          footer->has_rowindexstride() ? footer->rowindexstride() : 0},
+      types_{detail::convertToTypeProto(*footer)},
+      rawDataSize_{footer->has_rawdatasize() ? footer->rawdatasize() : 0},
+      checksumAlgorithm_{
+          footer->has_checksumalgorithm() ? footer->checksumalgorithm()
+                                          : proto::ChecksumAlgorithm::NULL_},
+      dwrfFooter_{footer} {
+  stripeCacheOffsets_.reserve(footer->stripecacheoffsets_size());
+  for (const auto& stripe : footer->stripes()) {
+    stripes_.emplace_back(stripe);
```

Comment:
This copy is not cheap with nested objects.  Can we make a zero copy wrapper view instead of copy the whole object tree?

Same thing applies to `metadata_` and `types_`.

Reply:
Hi @Yuhta, thanks for reviewing! Indeed it's non-optimal to copy all the objects.
I attempted to modify `TypeProto` to hold a pointer to the underlying `proto::Type` / `proto::orc::Type` and provides access to it's attributes. Please take a look if it's in the correct direction. If so, I will apply it to the rest of wrappers. Thanks.